node/node.go

```diff
@@ -468,11 +467,12 @@ func createTransport(
 		)
 	}
 
-	p2p.MultiplexTransportConnFilters(connFilters...)(transport)
-
-	// Limit the number of incoming connections.
-	max := config.P2P.MaxNumInboundPeers + len(splitAndTrimEmpty(config.P2P.UnconditionalPeerIDs, ",", " "))
-	p2p.MultiplexTransportMaxIncomingConnections(max)(transport)
+	transport := p2p.NewMConnTransport(
+		logger, nodeInfo, nodeKey.PrivKey, p2p.MConnConfig(config.P2P),
```

Comment:
Yeah, so I'm not really a fan of `SetLogger()`, because that means that 1) we don't have access to a logger during construction, and 2) there are race conditions where the component may be doing things in a goroutine before the logger gets set and thus panics due to `nil` logger. I had multiple problems with this because the logger gets set via `Peer.SetLogger`, which happens very late in the connection setup.

Personally I strongly prefer to just pass them to the constructor and get rid of `SetLogger`.

Reply:
> 1) we don't have access to a logger during construction

this shouldn't be a big problem. constructors need to be simple and don't do any heavy work anyway

> , and 2) there are race conditions where the component may be doing things in a goroutine before the logger gets set and thus panics due to nil logger.

- SetLogger must only be set right after constructor, not during lifetime of the object
- `log.NewNopLogger` must be used in constructor

personally I strongly prefer for logger not to be required field of the object API.

Reply:
Ok. I'd like to get @alexanderbez opinion here as well, but I'm fine with whatever you guys prefer.

>SetLogger must only be set right after constructor, not during lifetime of the object

What if we spawn a goroutine in the constructor?

Reply:
> What if we spawn a goroutine in the constructor?

then the logger is needed

Reply:
I prefer to have the logger as part of the object construction. This also bit me during the evidence reactor refactor where I was expecting to see logs from the evidence pool and wasn't -- I had to explicitly call `SetLogger` in the reactor constructor.

Reply:
Heh, this just bit me again in the existing P2P peer tests -- they call `SetLogger` to override the logger, but of course the concurrent peer code is also using/setting the logger, so the race detector rejects it. For the `SetLogger` pattern to be safe, the `logger` field must be protected by a mutex, which is both inconvenient and inefficient. Let's just not do `SetLogger`.